### ISSUE TITLE
dask-worker fail to inform scheduler of closing with UCX

### ISSUE THREAD

**Issue by pentschev** [dask/distributed#3576]
`dask-worker`s fail to inform scheduler of their closing when protocol used is UCX. I spent quite some time trying to debug this, but I got really confused with the async state in distributed. What I figured is that the TCP stream in Tornado will hit 

https://github.com/dask/distributed/blob/806a7e97285c5534b3e37e912cbc060a8036c56f/distributed/comm/tcp.py#L205

 which will raise an exception and break out of 

https://github.com/dask/distributed/blob/806a7e97285c5534b3e37e912cbc060a8036c56f/distributed/core.py#L456

 and cause the scheduler to close its end in 

https://github.com/dask/distributed/blob/806a7e97285c5534b3e37e912cbc060a8036c56f/distributed/core.py#L491

However, UCX isn't a stream and will not raise such an exception, causing the scheduler to remain awaiting 

https://github.com/dask/distributed/blob/806a7e97285c5534b3e37e912cbc060a8036c56f/distributed/core.py#L456

The issue above should be taken care by 

https://github.com/dask/distributed/blob/806a7e97285c5534b3e37e912cbc060a8036c56f/distributed/nanny.py#L674-L699

But as far as I could understand, the process exits before `do_stop` gets executed, which is where I think the bug is but I wasn't able to understand what of the async tasks causes the process to exit prematurely.

To reproduce this, one can open a scheduler and a worker as below and once the worker is connected press CTRL+C, which will not cause the scheduler to report that the worker has closed.

```
dask-scheduler --protocol ucx
dask-worker ucx://SERVER_IP:8786
```

@jacobtomlinson @pitrou appreciate if you guys can shed some light on this.

cc @quasiben

**Comment by jacobtomlinson** [comment]
You mention it is stuck awaiting `comm.read()`. Which is this function in UCX.

https://github.com/dask/distributed/blob/806a7e97285c5534b3e37e912cbc060a8036c56f/distributed/comm/ucx.py#L171-L209

It would be interesting to know where within that function it is awaiting and whether we can cancel the coroutine or raise a `CommClosedError` as a result.

**Comment by pentschev** [comment]
It is awaiting for a `self.ep.recv` (can't remember exactly which one, but I think the one on line 180), but this is the expected way for UCX, as it's not a stream as TCP is, so we must make sure one endpoint finishes its work and notifies the other end to close itself. In the case of `Nanny` this is handled by the block I mentioned earlier https://github.com/dask/distributed/blob/806a7e97285c5534b3e37e912cbc060a8036c56f/distributed/nanny.py#L674-L699 but that fails to complete as the process terminates before the async task has a chance to execute.

**Comment by jacobtomlinson** [comment]
What is terminating the process? A signal?

**Comment by quasiben** [comment]
I think in @pentschev's experiments, termination is done manually with crtl-c

**Comment by pentschev** [comment]
Sorry, I missed this earlier, and @quasiben is correct.

**Comment by mrocklin** [comment]
Checking in @pentschev @quasiben @jacobtomlinson did you all ever resolve what was happening here?  I'm curious if this issue is resolved, stale, or something that you're actively working on.

**Comment by quasiben** [comment]
It is not resolved -- perhaps stale for the moment but could be picked up again in 2 weeks or so.  @madsbk is working on a cleaner endpoint closing in ucx-py.

**Comment by jakirkham** [comment]
This should be solved by PR ( https://github.com/dask/distributed/pull/3747 ).

**Comment by quasiben** [comment]
closed by #3747
